### ISSUE TITLE
Account nonce not updating correctly

### ISSUE THREAD

**Issue by gspasov** [aeternity/elixir-node#314]
The Account nonce should be updated upon adding a transaction to the pool, because otherwise we cannot create more than one valid transaction in the block. At the moment the Account nonce is updated upon validation/mining of the block, so if I add more than one transaction the nonce will be equal for all of them, meaning only the first one will be mined, all other transactions will be considered invalid (nonce too small)

**Comment by d-velev** [comment]
This is a result of the automatic nonce set in transactions that we do to make transaction creation easier, the account nonce should only be updated when a transaction is mined and not when it's added to the pool.

**Comment by gspasov** [comment]
Transactions should be created automatically. Yes, we can ignore this if we do it manually, but this is only maybe useful in testing enviorment. In a real case scenario transaction data will be created automatically, you would only set who is the receiver, the amount you are sending and the fee of the transaction. Everything else should be automatic, eg. creating correct nonce, signing the transaction, getting your public key for the sender field.

**Comment by d-velev** [comment]
But we shouldn't change chainstate by adding a transaction to the pool, only when we mine one

**Comment by gspasov** [comment]
Then maybe there should be a state that holds the temporary new nonce of the sender's account

**Comment by cytadela8** [comment]
It would be good thing definitely to store such information somewhere and handle nonce's nicely. As a side note I would like to point out myetherwallet doesn't do that. (The nonce for new transaction is updated only after account state changes)

**Comment by cytadela8** [comment]
Probably last used nonce could be stored in Wallet.Worker or Aewallet.

**Comment by gspasov** [comment]
Now when I think about it this could lead to another problem. If let's say John creates 9 invalid transactions, but the 10th be a valid one. When the miner validates the transactions he'll get the 10th transaction with the nonce being 10, even though this will be John's first transaction.

Non the less, the way we have it now has another potential problem. Because the Pool doesn't validate the trasactions that nodes send to it, you can send any number of invalid transaction. A problem might occur if a malicious node sens 1_000_000+ invalid transactions just to slow down the miners in validating them, because the Pool will accept them if the `fee` is **high enough** and the `signature` is **correct**. But this doesn't mean that the transaction is valid whatsoever. This may potentialy brake the miner nodes by flooding them with milions of invalid transactions.

Maybe we should validate the transaction before adding them to the Pool? Suggestions? @cytadela8 @d-velev @thepiwo @rdimitrovbg @meivantodorov

**Comment by thepiwo** [comment]
> Now when I think about it, this could lead to another problem. If let's say John creates 9 invalid transactions, but the 10th be a valid one. When the miner validates the transactions he'll get the 10th transaction with the nonce being 10, even though this will be John's first transaction.

as intended

> you can send any number of invalid transaction

Is a late optimization problem, not related to validation. We need to block malicious peers in any case, they can send invalid blocks/provide invalid peers/lot of other things - same deal

> Maybe we should validate the transaction before adding them to the Pool?

we can't the pool is independent from state, some node may not know the validity in this point of time.

**Comment by cytadela8** [comment]
One potentially interesting approach for handling pool flooding would be to allow including invalid transactions in blocks if only signature and nonce are correct. Such transactions wouldn't execute, but fee would still be consumed.

**Comment by gspasov** [comment]
Isn't the fee consumed after the miner adds the transaction to the block? In that case no fee would be collected, because the miner won't add invalid transaction to the next block.

**Comment by thepiwo** [comment]
client-software should keep track of increasing nonces when multiple transactions should be included in one block.

The pool will at some point be sorted and trying to resolve dependencies within transactions.

Any kind of flooding is to be handled by the peer manager in general. Only valid transactions on chain-state will and should be included in blocks. Eventually fee wouldn't be able to consume anyways as is limits validity in some cases.

**Comment by thepiwo** [comment]
will close this as wontfix
